context.go

```diff
@@ -225,11 +289,59 @@ func (context *Context) Close() {
 
 // TotalRuntime returns the cumulated WAF runtime across various run calls within the same WAF context.
 // Returned time is in nanoseconds.
-func (context *Context) TotalRuntime() (overallRuntimeNs, internalRuntimeNs uint64) {
-	return context.totalOverallRuntimeNs.Load(), context.totalRuntimeNs.Load()
+// Deprecated: use Timings instead
+func (context *Context) TotalRuntime() (uint64, uint64) {
+	return uint64(context.stats.get("dd.appsec.waf.run") * time.Nanosecond), uint64(context.stats.get("dd.appsec.waf.duration"))
 }
 
 // TotalTimeouts returns the cumulated amount of WAF timeouts across various run calls within the same WAF context.
+// Deprecated: use Timings instead
 func (context *Context) TotalTimeouts() uint64 {
 	return context.timeoutCount.Load()
 }
+
+// Stats returns the cumulative time spent in various parts of the WAF, all in nanoseconds
+// and the timeout value used
+func (context *Context) Stats() map[string]time.Duration {
+	return context.stats.copy()
+}
+
+type metrics struct {
+	data  map[string]time.Duration
+	mutex sync.RWMutex
+}
+
+func (metrics *metrics) add(key string, duration time.Duration) {
+	metrics.mutex.Lock()
+	defer metrics.mutex.Unlock()
+	metrics.data[key] += duration
+}
+
+func (metrics *metrics) get(key string) time.Duration {
+	metrics.mutex.RLock()
+	defer metrics.mutex.RUnlock()
+	return metrics.data[key]
+}
+
+func (metrics *metrics) copy() map[string]time.Duration {
+	metrics.mutex.Lock()
+	defer metrics.mutex.Unlock()
+	copy := make(map[string]time.Duration, len(metrics.data))
+	for k, v := range metrics.data {
+		copy[k] = v
+	}
+	return copy
+}
+
+// merge merges the current metrics with the new run
+func (metrics *metrics) merge(other map[string]time.Duration) {
+	metrics.mutex.Lock()
+	defer metrics.mutex.Unlock()
+	for key, val := range other {
+		prev, ok := metrics.data[key]
+		if !ok {
+			prev = 0
+		}
+		metrics.data[key] = prev + val
```

Comment:
It feels like this could be delegated to `metrics.add`?

```suggestion
		metrics.add(key, val)
```